packages/lib/components/shared/note-screen-shared.ts

```diff
@@ -153,7 +153,7 @@ shared.saveNoteButton_press = async function(comp: BaseNoteScreenComponent, fold
 
 	// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Old code before rule was applied
 	const newState: any = {
-		lastSavedNote: { ...note },
+		lastSavedNote: { ...savedNote },
```

Comment:
To add additional explanation to what's provided in the pull request description, here's what I think was happening previously:
1. `saveNoteButton_press` is called and a mutex is acquired:
    https://github.com/laurent22/joplin/blob/388701015d72908b15496d658af8b2d3040e67cf/packages/lib/components/shared/note-screen-shared.ts#L101
2. The note is saved asynchronously:
    https://github.com/laurent22/joplin/blob/388701015d72908b15496d658af8b2d3040e67cf/packages/lib/components/shared/note-screen-shared.ts#L132
3. While the note is saving, the title or body is changed.
4. In `saveNoteButton_press`, the cached `note` is updated to include properties changed by the save logic:
    https://github.com/laurent22/joplin/blob/388701015d72908b15496d658af8b2d3040e67cf/packages/lib/components/shared/note-screen-shared.ts#L140
5. [**BUG part 1:**] The cached `note` is updated with the changes to the title/body made in step 3:
    https://github.com/laurent22/joplin/blob/388701015d72908b15496d658af8b2d3040e67cf/packages/lib/components/shared/note-screen-shared.ts#L150
6. [**BUG part 2**:] `lastSavedNote` is set to the **updated note**, which includes unsaved content:
    https://github.com/laurent22/joplin/blob/388701015d72908b15496d658af8b2d3040e67cf/packages/lib/components/shared/note-screen-shared.ts#L156
7. State is updated and the mutex is released.
8. No further edits are made.
9. The next time, `saveNoteButton_press` is called, `lastSavedNote` is compared with `note`. Because `lastSavedNote` includes the latest (unsaved) content from step 6, **the note isn't saved**.

**Question**:
With this change, is it possible for `state.lastSavedNote` to have only a subset of the properties of `state.note` (e.g. if `Note.save` returns a different set of properties from what's in `note`)?
   - If so, this might make [`BaseModel.diffObjectFields`](https://github.com/laurent22/joplin/blob/388701015d72908b15496d658af8b2d3040e67cf/packages/lib/components/shared/note-screen-shared.ts#L122) return a non-empty array, even if the note hasn't changed. This could lead to unnecessary calls to `Note.save` and an incorrect `fields` option parameter for `Note.save`.
   - One possible fix could be to declare a `const lastSavedNote = {...note, ...savedNote}` before reassigning `note.title`/`note.body`. Alternatively, if `savedNote` always includes `title` and `body` properties,
    
```suggestion
		lastSavedNote: { ...note, ...savedNote },
```

Reply:
I agree with your understanding of what you outlined is happening.

Regarding the question, that's a fair point, savedNote is actually set from the note details reloaded from the database rather than just the state before saving. I think your proposed change of setting lastSavedNote: { ...note, ...savedNote } sounds good. I can check if that still fixes the issue and update.

Just to confirm (as I've never seen this syntax before, just making guesses as to what it means), will that merge the values of the note and savedNote objects, taking preference to the values in savedNote where they overlap? Also, what does the ... signify?

Reply:
> Just to confirm (as I've never seen this syntax before, just making guesses as to what it means), will that merge the values of the note and savedNote objects, taking preference to the values in savedNote where they overlap? Also, what does the ... signify?

I'm linking to [the relevant MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#description).

Reply:
Yep, your proposed change still works